File: bot/models/security.py
from sqlalchemy import Column, Integer, String, DateTime, Boolean, ForeignKey, Text, Float
from datetime import datetime
from database.db import Base


class AuditLog(Base):
    __tablename__ = "audit_logs"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    # Org-scoped audit trail (nullable so legacy/system events still write).
    org_id = Column(String(36), nullable=True, index=True)
    # Agent-scoped events stamp the agent id here instead of overloading user_id.
    agent_id = Column(String(64), nullable=True, index=True)
    event_type = Column(
        String(50), nullable=False, index=True
    )  # login, swap, withdrawal, settings_change, 2fa_toggle, whitelist_change
    details = Column(Text, nullable=True)  # JSON details
    ip_address = Column(String(45), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)


class WithdrawalWhitelist(Base):
    __tablename__ = "withdrawal_whitelist"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    chain = Column(String(50), nullable=False)
    address = Column(String(255), nullable=False)
    label = Column(String(100), nullable=True)
    cooldown_until = Column(DateTime, nullable=True)  # 24h after creation
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)


class BackupCode(Base):
    __tablename__ = "backup_codes"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    code_hash = Column(String(128), nullable=False)
    is_used = Column(Boolean, default=False)
    used_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)


class SpendEvent(Base):
    """Durable record of USD outflow for spending-limit windows.

    SwapTransaction.from_amount_usd historically held token amounts, so limit
    windows are computed from this table instead — rows are only ever written
    with a real USD value, by the swap engine at submission time.
    """

    __tablename__ = "spend_events"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    amount_usd = Column(Float, nullable=False)
    kind = Column(String(20), default="swap")  # swap, withdrawal, ...
    swap_id = Column(Integer, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, index=True)

File: api-ts/src/db/schema/security.ts
import {
	boolean,
	index,
	integer,
	pgTable,
	serial,
	text,
	timestamp,
	uuid,
	varchar,
} from 'drizzle-orm/pg-core'

export const auditLogs = pgTable(
	'audit_logs',
	{
		id: serial('id').primaryKey(),
		userId: integer('user_id').notNull(),
		// Org-scoped audit trail: nullable so legacy/system events (userId 0,
		// no org) still write. Set for any org-mutating or money-path event so an
		// enterprise admin can query org-wide activity — the institutional gate.
		orgId: uuid('org_id'),
		// Agent-scoped events stamp the agent id here instead of overloading userId.
		agentId: varchar('agent_id', { length: 64 }),
		eventType: varchar('event_type', { length: 50 }).notNull(),
		details: text('details'),
		ipAddress: varchar('ip_address', { length: 45 }),
		createdAt: timestamp('created_at').defaultNow(),
		// Hash-chain (tamper evidence). Chained per-org ('global' chain for
		// org-less entries). Nullable so pre-existing rows (written before this
		// migration) don't need backfill — the chain simply starts fresh from the
		// first row that has these populated. See services/audit.ts for the
		// hashing + locking scheme.
		prevHash: varchar('prev_hash', { length: 64 }),
		entryHash: varchar('entry_hash', { length: 64 }),
	},
	(t) => ({
		orgCreatedIdx: index('audit_logs_org_created_idx').on(t.orgId, t.createdAt),
		agentCreatedIdx: index('audit_logs_agent_created_idx').on(t.agentId, t.createdAt),
	}),
)

export const backupCodes = pgTable('backup_codes', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull(),
	codeHash: varchar('code_hash', { length: 128 }).notNull(),
	isUsed: boolean('is_used').default(false),
	usedAt: timestamp('used_at'),
	createdAt: timestamp('created_at').defaultNow(),
})

export type AuditLog = typeof auditLogs.$inferSelect
export type NewAuditLog = typeof auditLogs.$inferInsert
export type BackupCode = typeof backupCodes.$inferSelect
export type NewBackupCode = typeof backupCodes.$inferInsert

export const withdrawalWhitelist = pgTable('withdrawal_whitelist', {
	id: serial('id').primaryKey(),
	userId: integer('user_id').notNull(),
	chain: varchar('chain', { length: 50 }).notNull(),
	address: varchar('address', { length: 255 }).notNull(),
	label: varchar('label', { length: 100 }),
	cooldownUntil: timestamp('cooldown_until'),
	isActive: boolean('is_active').default(true),
	createdAt: timestamp('created_at').defaultNow(),
})

export type WithdrawalWhitelist = typeof withdrawalWhitelist.$inferSelect
export type NewWithdrawalWhitelist = typeof withdrawalWhitelist.$inferInsert
